catalog/IndexScheme.hpp

```diff
@@ -0,0 +1,183 @@
+   *        Protocol Buffer form.
+   *
+   * @param proto The Protocol Buffer serialization of a Index Scheme,
+   *              previously produced by getProto().
+   * @return The deserialied index scheme object.
+   **/
+  static IndexScheme* ReconstructFromProto(const serialization::IndexScheme &proto);
+
+  /**
+   * @brief Check whether a serialization::IndexScheme is fully-formed and
+   *        all parts are valid.
+   *
+   * @param proto A serialized Protocol Buffer representation of a
+   *              IndexScheme, originally generated by getProto().
+   * @return Whether proto is fully-formed and valid.
+   **/
+  static bool ProtoIsValid(const serialization::IndexScheme &proto);
+
+  /**
+   * @brief Serialize the Index Scheme as Protocol Buffer.
+   *
+   * @return The Protocol Buffer representation of Index Scheme.
+   **/
+  serialization::IndexScheme getProto() const;
+
+  /**
+   * @brief Get the number of indices for the relation.
+   *
+   * @return The number of indices defined for the relation.
+   **/
+  inline const std::size_t getNumIndices() const {
+    return index_map_.size();
+  }
+
+  /**
+   * @brief Check whether an index with the given exists or not.
+   *
+   * @param index_name Name of the index to be checked.
+   * @return Whether the index exists or not.
+   **/
+  bool hasIndexWithName(const std::string &index_name) const {
+    return index_map_.find(index_name) != index_map_.end();
+  }
+
+  /**
+   * @brief Check whether an index with the given description
+   *        containing the same attribute id and index type
+   *        exists or not in the index map.
+   *
+   * @param index_descripton Index Description to check against.
+   * @return Whether a similar index description was already defined or not.
+   **/
+  bool hasIndexWithDescription(const IndexSubBlockDescription &index_description) const {
+    // Iterate through every vector of descriptions corresponding to all keys in the index map.
+    for (auto cit = index_map_.cbegin(); cit != index_map_.cend(); ++cit) {
+      const std::vector<IndexSubBlockDescription> &index_descriptions = cit->second;
+      for (auto index_description_it = index_descriptions.cbegin();
+           index_description_it != index_descriptions.cend();
+           ++index_description_it) {
+        // Check if the stored description matches as the given description.
+        if (areIndexDescriptionsSimilar(*index_description_it, index_description)) {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * @brief Check whether two index descriptions are similar or not.
+   *        Two index descriptions are similar if they have same
+   *        attributes ids and same index type.
+   *
+   * @param desc_a First index description.
+   * @param desc_b Second index description.
+   * @return Whether the two index_descriptions are similar or not.
+   **/
+  bool areIndexDescriptionsSimilar(const IndexSubBlockDescription &desc_a,
```

Comment:
Rename to `areSameIndexDescriptions`? I think `similar` is ambiguous.

Also, rename the parameters to `description_expected` and `description_checked`, respectively. The former is for the known description, and the latter for the one used for comparison.